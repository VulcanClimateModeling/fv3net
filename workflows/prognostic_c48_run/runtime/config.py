from typing import List, Optional
import dataclasses
import yaml
import f90nml

import dacite

from runtime.diagnostics.manager import (
    DiagnosticFileConfig,
    FortranFileConfig,
    get_chunks,
)
from runtime.steppers.nudging import NudgingConfig
from runtime.steppers.machine_learning import MachineLearningConfig

FV3CONFIG_FILENAME = "fv3config.yml"


@dataclasses.dataclass
class UserConfig:
    """The top-level object for python runtime configurations

    Attributes:
        diagnostics: list of diagnostic file configurations
        fortran_diagnostics: list of Fortran diagnostic outputs. Currently only used by
            post-processing and so only name and chunks items need to be specified.
        scikit_learn: a machine learning configuration
        nudging: nudge2fine configuration. Cannot be used if any scikit_learn model
            urls are specified.
        step_tendency_variables: variables to compute the tendencies of.
            These could in principle be inferred from the requested diagnostic
            names.
        step_storage_variables: variables to compute the storage of. Needed for certain
            diagnostics.
    """

    diagnostics: List[DiagnosticFileConfig]
    fortran_diagnostics: List[FortranFileConfig]
    scikit_learn: MachineLearningConfig = MachineLearningConfig()
    nudging: Optional[NudgingConfig] = None
    step_tendency_variables: List[str] = dataclasses.field(
        default_factory=lambda: list(
            ("specific_humidity", "air_temperature", "eastward_wind", "northward_wind",)
        )
    )
    step_storage_variables: List[str] = dataclasses.field(
        default_factory=lambda: list(("specific_humidity", "total_water"))
    )


def get_config() -> UserConfig:
    """Open the configurations for this run
    
    .. warning::
        Only valid at runtime
    """
    with open("fv3config.yml") as f:
        config = yaml.safe_load(f)
    return dacite.from_dict(UserConfig, config)


def get_namelist() -> f90nml.Namelist:
    """Open the fv3 namelist
    
    .. warning::
        Only valid at runtime
    """
    return f90nml.read("input.nml")


def write_chunks(config: UserConfig):
    """Given UserConfig, write chunks to 'chunks.yaml'"""
    diagnostic_file_configs = (
        config.fortran_diagnostics + config.diagnostics  # type: ignore
    )
    chunks = get_chunks(diagnostic_file_configs)
    with open("chunks.yaml", "w") as f:
        yaml.safe_dump(chunks, f)
