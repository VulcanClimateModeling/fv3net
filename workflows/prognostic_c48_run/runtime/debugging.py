import pdb
from mpi4py import MPI
import xarray as xr


def breakpoint(rank=0):
    if rank == MPI.COMM_WORLD.rank:
        pdb.set_trace()


def checkpoint(state, path, rank=0):
    if MPI.COMM_WORLD.rank == 0:
        variables = state.checkpoint()
        input_ = xr.Dataset({key: variables[key] for key in variables})
        input_.to_netcdf(path)
