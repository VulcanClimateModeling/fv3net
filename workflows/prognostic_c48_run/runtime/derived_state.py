import cftime
from typing import Mapping, Hashable
import xarray as xr

import fv3gfs.util
from vcm import DerivedMapping


class FV3StateMapper:
    """ A mapping interface for the FV3GFS getter.
        
    Maps variables to the common names used in shared functions.
    By default adds mapping {"lon": "longitude", "lat": "latitude"}
    """

    def __init__(self, getter, alternate_keys: Mapping[str, str] = None):
        self._getter = getter
        self._alternate_keys = alternate_keys or {"lon": "longitude", "lat": "latitude"}

    def __getitem__(self, key: str) -> xr.DataArray:
        if key == "time":
            time = self._getter.get_state(["time"])["time"]
            return xr.DataArray(time, dims=["time"], coords={"time": [time]})
        elif key == "latent_heat_flux":
            return self._getter.get_diagnostic_by_name("lhtfl").data_array
        elif key == "total_water":
            return self._total_water()
        else:
            if key in self._alternate_keys:
                key = self._alternate_keys[key]
            return self._getter.get_state([key])[key].data_array

    def _total_water(self):
        a = self._getter.get_tracer_metadata()
        water_species = [name for name in a if a[name]["is_water"]]
        return sum(self[name] for name in water_species)


class DerivedFV3State:
    """A uniform mapping-like interface to the FV3GFS model state
    
    This class wraps the fv3gfs getters with the FV3StateMapper, that always returns
    DataArray and has time as an attribute (since this isn't a DataArray).
    
    This insulates runfiles from the details of Quantity
    
    """

    def __init__(self, getter):
        """
        Args:
            getter: the fv3gfs object or a mock of it.
        """
        self._getter = getter
        self._mapper = DerivedMapping(FV3StateMapper(getter, alternate_keys=None))

    @property
    def time(self) -> cftime.DatetimeJulian:
        return self._getter.get_state(["time"])["time"]

    def __getitem__(self, key: Hashable) -> xr.DataArray:
        return self._mapper[key]

    def __setitem__(self, key: str, value: xr.DataArray):
        self._getter.set_state_mass_conserving(
            {key: fv3gfs.util.Quantity.from_data_array(value)}
        )

    def update(self, items: Mapping[Hashable, xr.DataArray]):
        """Update state from another mapping

        This may be faster than setting each item individually.
        
        Same as dict.update.
        """
        self._getter.set_state_mass_conserving(
            {
                key: fv3gfs.util.Quantity.from_data_array(value)
                for key, value in items.items()
            }
        )
