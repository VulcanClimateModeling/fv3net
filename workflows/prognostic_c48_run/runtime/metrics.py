import jsonschema
import json
import os

_metrics_file = os.path.join(
    os.path.abspath(os.path.dirname(__file__)), "metrics-schema.json"
)
with open(_metrics_file) as f:
    SCHEMA = json.load(f)


def validate(obj):
    return jsonschema.validate(obj, SCHEMA)
