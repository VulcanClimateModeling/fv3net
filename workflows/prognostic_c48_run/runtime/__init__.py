from .config import get_namelist, get_config
from .capture import capture_stream, capture_stream_mpi, capture_fv3gfs_funcs
from .diagnostics import (
    get_diagnostic_files,
    DiagnosticFile,
    compute_ml_diagnostics,
    compute_ml_momentum_diagnostics,
    compute_nudging_diagnostics,
    rename_diagnostics,
    default_diagnostics,
)
from .adapters import RenamingAdapter, MultiModelAdapter
from .derived_state import DerivedFV3State
from .nudging import (
    nudging_timescales_from_dict,
    setup_get_reference_state,
    get_nudging_tendency,
    set_state_sst_to_reference,
)

__version__ = "0.1.0"
