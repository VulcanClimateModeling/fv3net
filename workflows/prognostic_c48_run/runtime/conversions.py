import xarray as xr
import pace.util

from runtime.types import QuantityState


def _insert_units_if_necessary(da: xr.DataArray) -> xr.DataArray:
    if "units" not in da.attrs:
        return da.assign_attrs(units="")
    else:
        return da


def dataset_to_quantity_state(ds: xr.Dataset) -> QuantityState:
    quantity_state: QuantityState = {
        variable: pace.util.Quantity.from_data_array(
            _insert_units_if_necessary(ds[variable])
        )
        for variable in ds.data_vars
    }
    return quantity_state


def quantity_state_to_dataset(quantity_state: QuantityState) -> xr.Dataset:
    ds = xr.Dataset(
        {
            variable: quantity_state[variable].data_array
            for variable in quantity_state.keys()
        }
    )
    return ds
