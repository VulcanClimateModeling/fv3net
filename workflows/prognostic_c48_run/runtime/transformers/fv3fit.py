import dataclasses
from typing import Mapping, Iterable, Hashable, Sequence, Union

import xarray as xr
import fv3fit
from runtime.steppers.machine_learning import non_negative_sphum, MultiModelAdapter
from runtime.types import State
from runtime.names import SPHUM

__all__ = ["Config", "Adapter"]


@dataclasses.dataclass
class Config:
    """
    Attributes:
        url: Path to a model to-be-loaded, or sequence of paths.
        variables: Mapping from state names to name of corresponding tendency predicted
            by model. For example: {"air_temperature": "dQ1"}.
        limit_negative_humidity: if True, rescale tendencies to not allow specific
            humidity to become negative.
        online: if True, the ML predictions will be applied to model state.
    """

    url: Union[str, Sequence[str]]
    variables: Mapping[str, str]
    limit_negative_humidity: bool = True
    online: bool = True


@dataclasses.dataclass
class Adapter:
    config: Config
    timestep: float

    def __post_init__(self: "Adapter"):
        if isinstance(self.config.url, str):
            self.model = fv3fit.load(self.config.url)
        else:
            models = [fv3fit.load(url) for url in self.config.url]
            self.model = MultiModelAdapter(models)

    def predict(self, inputs: State) -> State:
        tendencies = self.model.predict(xr.Dataset(inputs))
        if self.config.limit_negative_humidity:
            dQ1, dQ2 = non_negative_sphum(
                inputs[SPHUM], tendencies["dQ1"], tendencies["dQ2"], self.timestep
            )
            tendencies.update({"dQ1": dQ1, "dQ2": dQ2})
        state_prediction: State = {}
        for variable_name, tendency_name in self.config.variables.items():
            with xr.set_options(keep_attrs=True):
                state_prediction[variable_name] = (
                    inputs[variable_name] + tendencies[tendency_name] * self.timestep
                )
        return state_prediction

    def apply(self, prediction: State, state: State):
        if self.config.online:
            state.update(prediction)

    def partial_fit(self, inputs: State, state: State):
        pass

    @property
    def input_variables(self) -> Iterable[Hashable]:
        return self.model.input_variables
