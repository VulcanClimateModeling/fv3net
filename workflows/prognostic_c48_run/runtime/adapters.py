from typing import Mapping, Set, Sequence, Hashable

import xarray as xr

from fv3fit._shared import Predictor

NameDict = Mapping[Hashable, Hashable]


def _invert_dict(d: Mapping) -> Mapping:
    return dict(zip(d.values(), d.keys()))


class RenamingAdapter:
    """Adapter object for renaming model variables

    Attributes:
        model: a model to rename
        rename_in: mapping from standard names to input names of model
        rename_out: mapping from standard names to the output names of model
    
    """

    def __init__(
        self, model: Predictor, rename_in: NameDict, rename_out: NameDict = None
    ):
        self.model = model
        self.rename_in = rename_in
        self.rename_out = {} if rename_out is None else rename_out

    def _rename(self, ds: xr.Dataset, rename: NameDict) -> xr.Dataset:

        all_names = set(ds.dims) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        redimed = ds.rename_dims(rename_restricted)

        all_names = set(ds.data_vars) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        return redimed.rename(rename_restricted)

    def _rename_inputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, self.rename_in)

    def _rename_outputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, _invert_dict(self.rename_out))

    @property
    def input_variables(self) -> Set[str]:
        invert_rename_in = _invert_dict(self.rename_in)
        return {invert_rename_in.get(var, var) for var in self.model.input_variables}

    def predict(self, arg: xr.Dataset) -> xr.Dataset:
        input_ = self._rename_inputs(arg)
        prediction = self.model.predict(input_)
        return self._rename_outputs(prediction)


class StackingAdapter:
    """Wrap a model to work with unstacked inputs
    
    Attributes
        model: a model to stack
        sample_dims: sample dimension name of the resulting stacked array
    
    """

    def __init__(self, model: Predictor, sample_dims: Sequence[str]):
        self.model = model
        self.sample_dims = sample_dims

    @property
    def input_variables(self) -> Set[str]:
        return set(self.model.input_variables)

    def predict(self, ds: xr.Dataset) -> xr.Dataset:
        self.model.predict_columnwise(ds, sample_dims=self.sample_dims)
