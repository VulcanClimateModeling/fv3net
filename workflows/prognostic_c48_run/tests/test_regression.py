from pathlib import Path
import json
import fv3config
import fv3fit
import runtime.metrics
import tempfile
import numpy as np
import pytest
import xarray as xr
import datetime
import yaml
import vcm.testing
from machine_learning_mocks import get_mock_predictor

import subprocess

BASE_FV3CONFIG_CACHE = Path("vcm-fv3config", "data")
IC_PATH = BASE_FV3CONFIG_CACHE.joinpath(
    "initial_conditions", "c12_restart_initial_conditions", "v1.0"
)
ORO_PATH = BASE_FV3CONFIG_CACHE.joinpath("orographic_data", "v1.0")
FORCING_PATH = BASE_FV3CONFIG_CACHE.joinpath("base_forcing", "v1.1")
LOG_PATH = "logs.txt"
STATISTICS_PATH = "statistics.txt"
PROFILES_PATH = "profiles.txt"
CHUNKS_PATH = "chunks.yaml"


class ConfigEnum:
    nudging = "nudging"
    predictor = "predictor"


default_fv3config = rf"""
data_table: default
diag_table: default
experiment_name: default_experiment
forcing: gs://{FORCING_PATH.as_posix()}
initial_conditions: gs://{IC_PATH.as_posix()}
orographic_forcing: gs://{ORO_PATH.as_posix()}
nudging: null
namelist:
  amip_interp_nml:
    data_set: reynolds_oi
    date_out_of_range: climo
    interp_oi_sst: true
    no_anom_sst: false
    use_ncep_ice: false
    use_ncep_sst: true
  atmos_model_nml:
    blocksize: 24
    chksum_debug: false
    dycore_only: false
    fdiag: 0.0
    fhmax: 1024.0
    fhmaxhf: -1.0
    fhout: 0.25
    fhouthf: 0.0
  cires_ugwp_nml:
    knob_ugwp_azdir:
    - 2
    - 4
    - 4
    - 4
    knob_ugwp_doaxyz: 1
    knob_ugwp_doheat: 1
    knob_ugwp_dokdis: 0
    knob_ugwp_effac:
    - 1
    - 1
    - 1
    - 1
    knob_ugwp_ndx4lh: 4
    knob_ugwp_solver: 2
    knob_ugwp_source:
    - 1
    - 1
    - 1
    - 0
    knob_ugwp_stoch:
    - 0
    - 0
    - 0
    - 0
    knob_ugwp_version: 0
    knob_ugwp_wvspec:
    - 1
    - 32
    - 32
    - 32
    launch_level: 55
  coupler_nml:
    atmos_nthreads: 1
    calendar: julian
    force_date_from_namelist: true
    current_date:
    - 2016
    - 8
    - 1
    - 0
    - 0
    - 0
    days: 0
    dt_atmos: 900
    dt_ocean: 900
    hours: 0
    memuse_verbose: true
    minutes: 30
    months: 0
    ncores_per_node: 32
    seconds: 0
    use_hyper_thread: true
  diag_manager_nml:
    prepend_date: false
  external_ic_nml:
    checker_tr: false
    filtered_terrain: true
    gfs_dwinds: true
    levp: 64
    nt_checker: 0
  fms_io_nml:
    checksum_required: false
    max_files_r: 100
    max_files_w: 100
  fms_nml:
    clock_grain: ROUTINE
    domains_stack_size: 3000000
    print_memory_usage: false
  fv_core_nml:
    a_imp: 1.0
    adjust_dry_mass: false
    beta: 0.0
    consv_am: false
    consv_te: 1.0
    d2_bg: 0.0
    d2_bg_k1: 0.16
    d2_bg_k2: 0.02
    d4_bg: 0.15
    d_con: 1.0
    d_ext: 0.0
    dddmp: 0.2
    delt_max: 0.002
    dnats: 1
    do_sat_adj: true
    do_vort_damp: true
    dwind_2d: false
    external_ic: false
    fill: true
    fv_debug: false
    fv_sg_adj: 900
    gfs_phil: false
    hord_dp: 6
    hord_mt: 6
    hord_tm: 6
    hord_tr: 8
    hord_vt: 6
    hydrostatic: false
    io_layout:
    - 1
    - 1
    k_split: 1
    ke_bg: 0.0
    kord_mt: 10
    kord_tm: -10
    kord_tr: 10
    kord_wz: 10
    layout:
    - 1
    - 1
    make_nh: false
    mountain: true
    n_split: 6
    n_sponge: 4
    na_init: 0
    ncep_ic: false
    nggps_ic: false
    no_dycore: false
    nord: 2
    npx: 13
    npy: 13
    npz: 63
    ntiles: 6
    nudge: false
    nudge_qv: true
    nwat: 6
    p_fac: 0.1
    phys_hydrostatic: false
    print_freq: 3
    range_warn: true
    reset_eta: false
    rf_cutoff: 800.0
    rf_fast: false
    tau: 5.0
    use_hydro_pressure: false
    vtdm4: 0.06
    warm_start: true
    z_tracer: true
  fv_grid_nml: {{}}
  gfdl_cloud_microphysics_nml:
    c_cracw: 0.8
    c_paut: 0.5
    c_pgacs: 0.01
    c_psaci: 0.05
    ccn_l: 300.0
    ccn_o: 100.0
    const_vg: false
    const_vi: false
    const_vr: false
    const_vs: false
    de_ice: false
    do_qa: true
    do_sedi_heat: false
    dw_land: 0.16
    dw_ocean: 0.1
    fast_sat_adj: true
    fix_negative: true
    icloud_f: 1
    mono_prof: true
    mp_time: 450.0
    prog_ccn: false
    qi0_crt: 8.0e-05
    qi_lim: 1.0
    ql_gen: 0.001
    ql_mlt: 0.001
    qs0_crt: 0.001
    rad_graupel: true
    rad_rain: true
    rad_snow: true
    rh_inc: 0.3
    rh_inr: 0.3
    rh_ins: 0.3
    rthresh: 1.0e-05
    sedi_transport: false
    tau_g2v: 900.0
    tau_i2s: 1000.0
    tau_l2v:
    - 225.0
    tau_v2l: 150.0
    use_ccn: true
    use_ppm: false
    vg_max: 12.0
    vi_max: 1.0
    vr_max: 12.0
    vs_max: 2.0
    z_slope_ice: true
    z_slope_liq: true
  gfs_physics_nml:
    cal_pre: false
    cdmbgwd:
    - 3.5
    - 0.25
    cnvcld: false
    cnvgwd: true
    debug: false
    dspheat: true
    fhcyc: 24.0
    fhlwr: 3600.0
    fhswr: 3600.0
    fhzero: 0.25
    hybedmf: true
    iaer: 111
    ialb: 1
    ico2: 2
    iems: 1
    imfdeepcnv: 2
    imfshalcnv: 2
    imp_physics: 11
    isol: 2
    isot: 1
    isubc_lw: 2
    isubc_sw: 2
    ivegsrc: 1
    ldiag3d: true
    lwhtr: true
    ncld: 5
    nst_anl: true
    pdfcld: false
    pre_rad: false
    prslrd0: 0.0
    random_clds: false
    redrag: true
    shal_cnv: true
    swhtr: true
    trans_trac: true
    use_ufo: true
  interpolator_nml:
    interp_method: conserve_great_circle
  nam_stochy:
    lat_s: 96
    lon_s: 192
    ntrunc: 94
  namsfc:
    fabsl: 99999
    faisl: 99999
    faiss: 99999
    fnabsc: grb/global_mxsnoalb.uariz.t1534.3072.1536.rg.grb
    fnacna: ''
    fnaisc: grb/CFSR.SEAICE.1982.2012.monthly.clim.grb
    fnalbc: grb/global_snowfree_albedo.bosu.t1534.3072.1536.rg.grb
    fnalbc2: grb/global_albedo4.1x1.grb
    fnglac: grb/global_glacier.2x2.grb
    fnmskh: grb/seaice_newland.grb
    fnmxic: grb/global_maxice.2x2.grb
    fnslpc: grb/global_slope.1x1.grb
    fnsmcc: grb/global_soilmgldas.t1534.3072.1536.grb
    fnsnoa: ''
    fnsnoc: grb/global_snoclim.1.875.grb
    fnsotc: grb/global_soiltype.statsgo.t1534.3072.1536.rg.grb
    fntg3c: grb/global_tg3clim.2.6x1.5.grb
    fntsfa: ''
    fntsfc: grb/RTGSST.1982.2012.monthly.clim.grb
    fnvegc: grb/global_vegfrac.0.144.decpercent.grb
    fnvetc: grb/global_vegtype.igbp.t1534.3072.1536.rg.grb
    fnvmnc: grb/global_shdmin.0.144x0.144.grb
    fnvmxc: grb/global_shdmax.0.144x0.144.grb
    fnzorc: igbp
    fsicl: 99999
    fsics: 99999
    fslpl: 99999
    fsmcl:
    - 99999
    - 99999
    - 99999
    fsnol: 99999
    fsnos: 99999
    fsotl: 99999
    ftsfl: 99999
    ftsfs: 90
    fvetl: 99999
    fvmnl: 99999
    fvmxl: 99999
    ldebug: false
"""

# Necessary to know the number of restart timestamp folders to generate in fixture
START_TIME = [2016, 8, 1, 0, 0, 0]
TIMESTEP_MINUTES = 15
NUM_NUDGING_TIMESTEPS = 2
RUNTIME_MINUTES = TIMESTEP_MINUTES * NUM_NUDGING_TIMESTEPS
TIME_FMT = "%Y%m%d.%H%M%S"
RUNTIME = {"days": 0, "months": 0, "hours": 0, "minutes": RUNTIME_MINUTES, "seconds": 0}


def run_native(config, rundir):
    with tempfile.NamedTemporaryFile("w") as f:
        yaml.safe_dump(config, f)
        fv3_script = Path(__file__).parent.parent.joinpath("runfv3").as_posix()
        subprocess.check_call([fv3_script, "create", rundir, f.name])
        subprocess.check_call([fv3_script, "append", rundir])


def assets_from_initial_condition_dir(dir_: str):
    start = datetime.datetime(*START_TIME)  # type: ignore
    delta_t = datetime.timedelta(minutes=TIMESTEP_MINUTES)
    assets = []
    for i in range(NUM_NUDGING_TIMESTEPS + 1):
        timestamp = (start + i * delta_t).strftime(TIME_FMT)

        for tile in range(1, 7):
            for category in [
                "fv_core.res",
                "fv_srf_wnd.res",
                "fv_tracer.res",
                "phy_data",
                "sfc_data",
            ]:
                assets.append(
                    fv3config.get_asset_dict(
                        dir_,
                        f"{category}.tile{tile}.nc",
                        target_location=timestamp,
                        target_name=f"{category}.tile{tile}.nc",
                    )
                )
    return assets


def _get_nudging_config(config_yaml: str, timestamp_dir: str):
    config = yaml.safe_load(config_yaml)
    coupler_nml = config["namelist"]["coupler_nml"]
    coupler_nml["current_date"] = START_TIME
    coupler_nml.update(RUNTIME)

    config["nudging"] = {
        "restarts_path": ".",
        "timescale_hours": {"air_temperature": 3.0, "specific_humidity": 3.0},
    }

    config.setdefault("patch_files", []).extend(
        assets_from_initial_condition_dir(timestamp_dir)
    )
    if coupler_nml["dt_atmos"] // 60 != TIMESTEP_MINUTES:
        raise ValueError(
            "Model timestep in default_fv3config not aligned"
            " with specified module's TIMESTEP_MINUTES variable."
        )

    return config


def get_nudging_config():
    config = _get_nudging_config(default_fv3config, "gs://" + IC_PATH.as_posix())
    config["diagnostics"] = [
        {
            "name": "diags.zarr",
            "times": {"kind": "interval", "frequency": 900, "times": None},
            "variables": [
                "air_temperature_reference",
                "air_temperature_tendency_due_to_nudging",
                "area",
                "cnvprcp_after_physics",
                "cnvprcp_after_python",
                "evaporation",
                "column_heating_due_to_nudging",
                "net_moistening_due_to_nudging",
                "physics_precip",
                "specific_humidity_reference",
                "specific_humidity_tendency_due_to_nudging",
                "storage_of_mass_due_to_fv3_physics",
                "storage_of_mass_due_to_python",
                "storage_of_specific_humidity_path_due_to_fv3_physics",
                "storage_of_specific_humidity_path_due_to_microphysics",
                "storage_of_specific_humidity_path_due_to_python",
                "storage_of_total_water_path_due_to_fv3_physics",
                "storage_of_total_water_path_due_to_python",
                "surface_temperature_reference",
                "tendency_of_air_temperature_due_to_fv3_physics",
                "tendency_of_air_temperature_due_to_python",
                "tendency_of_eastward_wind_due_to_fv3_physics",
                "tendency_of_eastward_wind_due_to_python",
                "tendency_of_northward_wind_due_to_fv3_physics",
                "tendency_of_northward_wind_due_to_python",
                "tendency_of_specific_humidity_due_to_fv3_physics",
                "tendency_of_specific_humidity_due_to_python",
                "total_precip_after_physics",
                "total_precipitation_rate",
                "water_vapor_path",
            ],
        }
    ]
    config["fortran_diagnostics"] = []
    return config


def get_ml_config(model_path):
    config = yaml.safe_load(default_fv3config)
    config["diagnostics"] = [
        {
            "name": "diags.zarr",
            "times": {"kind": "interval", "frequency": 900, "times": None},
            "variables": [
                "air_temperature",
                "area",
                "cnvprcp_after_physics",
                "cnvprcp_after_python",
                "column_integrated_dQ1_change_non_neg_sphum_constraint",
                "column_integrated_dQ2_change_non_neg_sphum_constraint",
                "column_integrated_dQu_stress",
                "column_integrated_dQv_stress",
                "dQ1",
                "dQ2",
                "dQu",
                "dQv",
                "evaporation",
                "column_heating_due_to_machine_learning",
                "net_moistening_due_to_machine_learning",
                "physics_precip",
                "pressure_thickness_of_atmospheric_layer",
                "specific_humidity",
                "specific_humidity_limiter_active",
                "storage_of_mass_due_to_fv3_physics",
                "storage_of_mass_due_to_python",
                "storage_of_specific_humidity_path_due_to_fv3_physics",
                "storage_of_specific_humidity_path_due_to_microphysics",
                "storage_of_specific_humidity_path_due_to_python",
                "storage_of_total_water_path_due_to_fv3_physics",
                "storage_of_total_water_path_due_to_python",
                "tendency_of_air_temperature_due_to_fv3_physics",
                "tendency_of_air_temperature_due_to_python",
                "tendency_of_eastward_wind_due_to_fv3_physics",
                "tendency_of_eastward_wind_due_to_python",
                "tendency_of_northward_wind_due_to_fv3_physics",
                "tendency_of_northward_wind_due_to_python",
                "tendency_of_specific_humidity_due_to_fv3_physics",
                "tendency_of_specific_humidity_due_to_python",
                "total_precip_after_physics",
                "total_precipitation_rate",
                "water_vapor_path",
            ],
        }
    ]
    config["fortran_diagnostics"] = []
    config["scikit_learn"] = {"model": [model_path]}
    # use local paths in prognostic_run image. fv3config
    # downloads data. We should change this once the fixes in
    # https://github.com/VulcanClimateModeling/fv3gfs-python/pull/78 propagates
    # into the prognostic_run image
    return config


@pytest.fixture(scope="module", params=[ConfigEnum.predictor, ConfigEnum.nudging])
def configuration(request):
    return request.param


@pytest.fixture(scope="module")
def completed_rundir(configuration, tmpdir_factory):

    model_path = str(tmpdir_factory.mktemp("model"))

    if configuration == ConfigEnum.predictor:
        model = get_mock_predictor()
        fv3fit.dump(model, str(model_path))
        config = get_ml_config(model_path)
    elif configuration == ConfigEnum.nudging:
        config = get_nudging_config()
    else:
        raise NotImplementedError()

    rundir = tmpdir_factory.mktemp("rundir").join("subdir")
    run_native(config, str(rundir))
    return rundir


@pytest.fixture()
def completed_segment(completed_rundir):
    return completed_rundir.join("artifacts").join("20160801.000000")


def test_fv3run_checksum_restarts(completed_segment, regtest):
    """Please do not add more test cases here as this test slows image build time.
    Additional Predictor model types and configurations should be tested against
    the base class in the fv3fit test suite.
    """
    fv_core = completed_segment.join("RESTART").join("fv_core.res.tile1.nc")
    print(fv_core.computehash(), file=regtest)


@pytest.mark.parametrize("path", [LOG_PATH, STATISTICS_PATH, PROFILES_PATH])
def test_fv3run_logs_present(completed_segment, path):
    assert completed_segment.join(path).exists()


def test_chunks_present(completed_segment):
    assert completed_segment.join(CHUNKS_PATH).exists()


def test_fv3run_diagnostic_outputs_check_variables(regtest, completed_rundir):
    """Please do not add more test cases here as this test slows image build time.
    Additional Predictor model types and configurations should be tested against
    the base class in the fv3fit test suite.
    """
    diagnostics = xr.open_zarr(str(completed_rundir.join("diags.zarr")))
    for variable in sorted(diagnostics):
        assert np.sum(np.isnan(diagnostics[variable].values)) == 0
        checksum = vcm.testing.checksum_dataarray(diagnostics[variable])
        print(f"{variable}: " + checksum, file=regtest)


def test_fv3run_diagnostic_outputs_schema(regtest, completed_rundir):
    diagnostics = xr.open_zarr(str(completed_rundir.join("diags.zarr")))
    diagnostics.info(regtest)


def test_fv3run_python_mass_conserving(completed_segment, configuration):
    if configuration == ConfigEnum.nudging:
        pytest.skip()

    path = str(completed_segment.join(STATISTICS_PATH))

    # read python mass conservation info
    with open(path) as f:
        lines = f.readlines()

    assert len(lines) > 0
    for metric in lines:
        obj = json.loads(metric)
        runtime.metrics.validate(obj)

        np.testing.assert_allclose(
            obj["storage_of_mass_due_to_python"],
            obj["storage_of_total_water_path_due_to_python"] * 9.81,
            rtol=0.003,
            atol=1e-4 / 86400,
        )


def test_fv3run_vertical_profile_statistics(completed_segment, configuration):
    if configuration == ConfigEnum.nudging:
        # no specific humidity limiter for nudging run
        pytest.skip()
    path = str(completed_segment.join(PROFILES_PATH))
    npz = yaml.safe_load(default_fv3config)["namelist"]["fv_core_nml"]["npz"]
    with open(path) as f:
        lines = f.readlines()

    for line in lines:
        profiles = json.loads(line)
        assert "time" in profiles
        assert len(profiles["specific_humidity_limiter_active_global_sum"]) == npz
