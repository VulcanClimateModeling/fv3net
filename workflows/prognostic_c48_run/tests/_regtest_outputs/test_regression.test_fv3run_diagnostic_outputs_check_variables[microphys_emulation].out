area: 51eca4ed49c60e8af827ffedcb42aa5e
cnvprcp_after_physics: 3c1d378be93048aa4c616310307c265d
evaporation: 7a780e75faa6c33fce1d6d69ad33ba3d
physics_precip: 6ced5b82629c88de88a6558d970ffc4e
storage_of_internal_energy_path_due_to_fv3_physics: 31afb571e6dc1c61dedf60c169615913
storage_of_mass_due_to_fv3_physics: 20db082e517601aa10da0f7123138e13
storage_of_specific_humidity_path_due_to_fv3_physics: 7c93387a3328bd6a6c88e0be0fa82e2c
storage_of_specific_humidity_path_due_to_microphysics: b702a44f1524da4b3e550e2c00d159c0
storage_of_total_water_path_due_to_fv3_physics: c1e863b81932ec1437a86858559c9aef
tendency_of_air_temperature_due_to_fv3_physics: 06e9680998b3ad5abced79f85f8d4957
tendency_of_eastward_wind_due_to_fv3_physics: 50af546fe6528c20f48580d3a2ca7673
tendency_of_internal_energy_due_to_fv3_physics: f20081418da59f099034144958938373
tendency_of_northward_wind_due_to_fv3_physics: dfe8bbe158ba47446bd77be2ea899e35
tendency_of_specific_humidity_due_to_fv3_physics: 39199f8b3d02e5db3e6b7f13edc8a2f4
total_precip_after_physics: ea1a76fb35674ecd53aaf1f65e96de94
total_precipitation_rate: 33d3bc207c3219bf66e573d56645b092
water_vapor_path: 6f23ca222f7aafdc1aeeaf68a8ba15b1
